heron/ckptmgr/src/java/com/twitter/heron/ckptmgr/CheckpointManagerServer.java

```diff
@@ -0,0 +1,297 @@
+import com.twitter.heron.common.network.HeronSocketOptions;
+import com.twitter.heron.common.network.REQID;
+import com.twitter.heron.proto.ckptmgr.CheckpointManager;
+import com.twitter.heron.proto.system.Common;
+
+public class CheckpointManagerServer extends HeronServer {
+  private static final Logger LOG = Logger.getLogger(CheckpointManagerServer.class.getName());
+
+  private final String topologyName;
+  private final String topologyId;
+  private final String checkpointMgrId;
+  private final IBackend checkpointsBackend;
+
+  private SocketChannel connection;
+
+  public CheckpointManagerServer(
+      String topologyName, String topologyId, String checkpointMgrId,
+      IBackend checkpointsBackend, NIOLooper s, String host,
+      int port, HeronSocketOptions options) {
+    super(s, host, port, options);
+
+    this.topologyName = topologyName;
+    this.topologyId = topologyId;
+    this.checkpointMgrId = checkpointMgrId;
+    this.checkpointsBackend = checkpointsBackend;
+
+    this.connection = null;
+
+    registerInitialization();
+  }
+
+  private void registerInitialization() {
+    registerOnRequest(CheckpointManager.RegisterStMgrRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.RegisterTMasterRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.SaveInstanceStateRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.GetInstanceStateRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.CleanStatefulCheckpointRequest.newBuilder());
+  }
+
+  @Override
+  public void onConnect(SocketChannel channel) {
+    LOG.info("Got a new connection from host:port "
+        + channel.socket().getRemoteSocketAddress());
+  }
+
+  @Override
+  public void onRequest(REQID rid, SocketChannel channel, Message request) {
+    if (request instanceof CheckpointManager.RegisterStMgrRequest) {
+      handleStMgrRegisterRequest(rid, channel, (CheckpointManager.RegisterStMgrRequest) request);
+    } else if (request instanceof CheckpointManager.RegisterTMasterRequest) {
+      handleTMasterRegisterRequest(rid, channel,
+                                   (CheckpointManager.RegisterTMasterRequest) request);
+    } else if (request instanceof CheckpointManager.SaveInstanceStateRequest) {
+      handleSaveInstanceStateRequest(
+          rid, channel, (CheckpointManager.SaveInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.GetInstanceStateRequest) {
+      handleGetInstanceStateRequest(
+          rid, channel, (CheckpointManager.GetInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.CleanStatefulCheckpointRequest) {
+      handleCleanStatefulCheckpointRequest(
+          rid, channel, (CheckpointManager.CleanStatefulCheckpointRequest) request);
+    } else {
+      LOG.severe("Unknown kind of request: " + request.getClass().getName());
+    }
+  }
+
+  protected void handleCleanStatefulCheckpointRequest(
```

Comment:
By following an OO approach using a handler interface (versus the procedural approach) we can support the following:

- ability to share common code in abstract or concrete superclasses (
- ability to use the decorator or filter pattern for common handler actions
- more modular unit tests and test classes
- smaller classes and methods
- easier to understand code (the RPC function name is the class name)

If you'd like to not expose the handler implementations to other classes you could use access level modifiers on the classes, or the methods, or better yet, package classes in a way that is specific to their function or preferred visibility.